osu.Game/Beatmaps/Beatmap.cs

```diff
@@ -50,7 +50,15 @@ public class Beatmap<T> : IBeatmap<T>
 
         IBeatmap IBeatmap.Clone() => Clone();
 
-        public Beatmap<T> Clone() => (Beatmap<T>)MemberwiseClone();
+        public Beatmap<T> Clone()
+        {
+            var clone = (Beatmap<T>)MemberwiseClone();
+
+            clone.ControlPointInfo = ControlPointInfo.CreateCopy();
+            // todo: deep clone other elements as required.
```

Comment:
Not sure about this... The method is listed as creating a shallow clone:

![image](https://user-images.githubusercontent.com/1329837/104673209-9a0e4e00-5724-11eb-8690-b50aa1758a0c.png)

Also kinda scary from a diffcalc/beatmap conversion performance perspective (beatmap conversion clones the beatmap, potentially 160+ times depending on the ruleset (mania)).

Reply:
This can be reverted if required (was a change made during review).

Reply:
I guess I'd be okay with reverting. It would probably mean that control points would have to be globally mutable again which I find super unfortunate but given there's a reason for it...

Reply:
That kinda goes with anything and everything though, perhaps a good path is to make beatmaps readonly but I'm not sure how far that hole goes.